File: fibonacci_memo.py
from typing import Dict
import unittest

def fib(n: int, memo: Dict[int, int] = None) -> int:
    """
    Calculates the nth Fibonacci number using recursion and memoization.
    fib(0) = 0, fib(1) = 1
    """
    if memo is None:
        memo = {0: 0, 1: 1}
    
    if n in memo:
        return memo[n]
    
    if n < 0:
        raise ValueError("n must be a non-negative integer")
        
    memo[n] = fib(n - 1, memo) + fib(n - 2, memo)
    return memo[n]

class TestFibonacci(unittest.TestCase):
    def test_base_cases(self):
        self.assertEqual(fib(0), 0)
        self.assertEqual(fib(1), 1)

    def test_values(self):
        self.assertEqual(fib(2), 1)
        self.assertEqual(fib(3), 2)
        self.assertEqual(fib(5), 5)
        self.assertEqual(fib(10), 55)
        self.assertEqual(fib(50), 12586269025)

if __name__ == "__main__":
    unittest.main()

File: fibonacci_memo.js
/**
 * Return the n-th Fibonacci number using memoization.
 * Time complexity: O(n)
 * Space complexity: O(n)
 * @param {number} n - Non-negative integer index
 * @returns {number}
 */
export function fib(n) {
  if (!Number.isInteger(n) || n < 0) {
    throw new RangeError('n must be a non-negative integer');
  }

  const memo = new Map([[0, 0], [1, 1]]);

  function compute(index) {
    if (memo.has(index)) {
      return memo.get(index);
    }

    const value = compute(index - 1) + compute(index - 2);
    memo.set(index, value);
    return value;
  }

  return compute(n);
}

// Optional: simple test when run directly
if (typeof require !== 'undefined' && require.main === module) {
  const assert = require('node:assert/strict');
  assert.equal(fib(0), 0);
  assert.equal(fib(1), 1);
  assert.equal(fib(10), 55);
  assert.equal(fib(50), 12586269025);
  assert.throws(() => fib(-1), RangeError);
  console.log('All tests passed');
}